// _@ali_uniform-react-components@0.0.35@@ali/uniform-react-components/lib/FormArr/index.test.js
// test FormArr
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _form = require('antd/lib/form');

var _form2 = _interopRequireDefault(_form);

var _button = require('antd/lib/button');

var _button2 = _interopRequireDefault(_button);

var _input = require('antd/lib/input');

var _input2 = _interopRequireDefault(_input);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('antd/lib/form/style');

require('antd/lib/button/style');

require('antd/lib/input/style');

var _index = require('./index');

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var FormItem = _form2.default.Item;

var EditForm = function (_React$Component) {
  _inherits(EditForm, _React$Component);

  function EditForm() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, EditForm);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = EditForm.__proto__ || Object.getPrototypeOf(EditForm)).call.apply(_ref, [this].concat(args))), _this), _this.save = function () {
      _this.props.form.validateFieldsAndScroll(function (err, values) {
        if (!err) {
          console.log('values: ', values);
        }
      });
    }, _this.cancel = function () {
      _this.props.onCancel();
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(EditForm, [{
    key: 'render',
    value: function render() {
      var data = this.props.data || {};
      var getFieldDecorator = this.props.form.getFieldDecorator;

      var formItemLayout = {
        labelCol: { span: 5 },
        wrapperCol: { span: 15 }
      };

      return _react2.default.createElement(
        _form2.default,
        { className: 'short-link-long-editformComponent' },
        _react2.default.createElement(
          _index2.default,
          {
            form: this.props.form,
            model: 'name'
          },
          _react2.default.createElement(
            'div',
            null,
            _react2.default.createElement(_input2.default, { style: { width: 300, marginLeft: '20.8%', marginBottom: 20 } })
          )
        ),
        _react2.default.createElement(
          'div',
          { className: 'modal-footer' },
          _react2.default.createElement(
            _button2.default,
            { type: 'primary', size: 'large', onClick: this.save },
            '\u4FDD\u5B58'
          ),
          _react2.default.createElement(
            _button2.default,
            { size: 'large', style: { marginLeft: 16 }, onClick: this.cancel },
            '\u53D6\u6D88'
          )
        )
      );
    }
  }]);

  return EditForm;
}(_react2.default.Component);

EditForm.displayName = 'EditForm';
var WrappeEditForm = _form2.default.create()(EditForm);
exports.default = WrappeEditForm;


ReactDOM.render(_react2.default.createElement(WrappeEditForm, null), document.getElementById('app'));
module.exports = exports['default'];
