// _@ali_uniform-react-components@0.0.35@@ali/uniform-react-components/src/Switch/index.js
'use strict';

import React from 'react';
import { Switch as SwitchAntd } from 'antd';
import './index.less';

class Switch extends React.Component {
  constructor(props) {
    super(props);
    const checked = (this.props.value == 1) ? true : false;
    let defaultProps = {
      checkedChildren: '开',
      unCheckedChildren: '关'
    }
    this.state = {
      ...defaultProps,
      ...this.props,
      checked
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps) {
      const value = nextProps.value;
      const checked = (value == 1) ? true : false;
      this.setState({ checked });
    }
  }

  handleChange = (checked) => {
    if (! ('value' in this.props)) {
      this.setState({ checked });
    }
    const value  = checked ? 1 : 0;
    const onChange = this.props.onChange;
    onChange && onChange(value);
  }

  render() {
    return (
      <SwitchAntd {...this.state} onChange={this.handleChange} />
    );
  }
}

Switch.displayName = 'Switch';

export default Switch;
